design-documents/service-communication-exception-handling.md

```diff
@@ -0,0 +1,107 @@
+## Exception handling in service communication
+
+### Problem
+
+We need to handle network exceptions that may happen in communication between services.
+
+This proposal aims to answer the following questions:
+* Should we use PHP exception class name in API responses or exception codes
+* What should we do with network exceptions (throw or suppress)
+* How do we limit amount of services we need to redeploy when one of the services start to return more exceptions
+
+As we are leaning towards using REST for communication between services, so this proposal will focus on REST.
+
+### Handling exceptions in API best practices
+
+For REST APIs the common practice is to incorporate response code in HTTP header. However, it's not enough for a client to do optimal exception handling as it's not granular enough. This is why most APIs include error message and exception code transferred as a part of message.
+
+#### Examples
+
+Twitter
+
+```
+{
+    "errors": [
+        {
+            "code": 215,
+            "message": "Bad Authentication data."
+        }
+    ]
+}
+```
+
+Facebook
+
+```
+{
+    "error": {
+        "message": "An active access token must be used to query information about the current user.",
+        "type": "OAuthException",
+        "code": 2500,
+        "fbtrace_id": "DzkTMkgIA7V"
+    }
+}
+```
+
+Twilio
+
+```
+<?xml version='1.0' encoding='UTF-8'?>
+<TwilioResponse>
+    <RestException>
+        <Code>20404</Code>
+        <Message>The requested resource /2010-04-01/Accounts/1234/IncomingPhoneNumbers/1234 was not found</Message>
+        <MoreInfo>https://www.twilio.com/docs/errors/20404</MoreInfo>
+        <Status>404</Status>
+    </RestException>
+</TwilioResponse>
+```
+
+### Exception class name or class name
+
+Existing REST API returns only error message, so we need to add exception class name or error code in addition to HTTP response code. This would allow different consumers of the service handle errors or understand what exception need to be handled or which exception type need to be thrown.
```

Comment:
Please describe specific use cases when having existing HTTP status codes + error messages is not enough?

BTW, in current implementation codes are not implemented on purpose because we have a way to distinguish exceptions on the client based on the message template (which has placeholders instead of real variable names). So there is a way to distinguish different exception instances, however there is no way to distinguish exception classes. It is also possible to distinguish specific exceptions like Authorization, Authentication, Not Found, Internal Server error and maybe some others.

Reply:
Having a mapping of message texts to exceptions is not convenient. Also the fact that messages can be used to understand what type of exception is thrown might be confusing, having error codes is common practice. I personally didn't know that you can depend on message text.

Reply:
This is legacy architecture. We need to weight pros and cons of introducing alternative.